src/Illuminate/Database/Query/Builder.php

```diff
@@ -1395,17 +1395,19 @@ protected function runSelect()
      * @param  int  $perPage
      * @param  array  $columns
      * @param  string  $pageName
+     * @param  int|null  $page
      * @return \Illuminate\Contracts\Pagination\LengthAwarePaginator
      */
-    public function paginate($perPage = 15, $columns = ['*'], $pageName = 'page')
+    public function paginate($perPage = 15, $columns = ['*'], $pageName = 'page', $page = null)
     {
-        $page = Paginator::resolveCurrentPage($pageName);
-
         $total = $this->getCountForPagination($columns);
 
-        $results = $this->forPage($page, $perPage)->get($columns);
+        $this->forPage(
+            $page = $page ?: Paginator::resolveCurrentPage($pageName),
```

Comment:
Can you move this assignment back to where it was and revert all changes to the way the forPage function is called.